lib/core/base/virtual-node.js

```diff
@@ -0,0 +1,79 @@
+// class is unused in the file...
+// eslint-disable-next-line no-unused-vars
+class VirtualNode {
+	/**
+	 * Wrap the real node and provide list of the flattened children
+	 *
+	 * @param node {Node} - the node in question
+	 * @param shadowId {String} - the ID of the shadow DOM to which this node belongs
+	 */
+	constructor(node, shadowId) {
+		this.shadowId = shadowId;
+		this.children = [];
+		this.actualNode = node;
+
+		this._isHidden = null; // will be populated by axe.utils.isHidden
+		this._cache = {};
+
+		// abstract Node and Element APIs so we can run axe in DOM-less
+		// environments. these are static properties in the assumption
+		// that axe does not change any of them while it runs.
+		this.elementNodeType = node.nodeType;
+		this.elementNodeName = node.nodeName.toLowerCase();
```

Comment:
IIRC @jkodu said there was some (undocumented) convention in this codebase to always use upper-case node names. I don't know if this matters (or is true), but if is, we should follow convention.

Reply:
Wilco asked for this so I'm going to assume the undocumented convention is now set to lowerCase. https://github.com/dequelabs/axe-core/pull/1562#discussion_r283401772